[SILOptimizer] Don't diagnose infinite recursion if a branch terminates the program

harlanhaskins opened the pull request:
This patch augments the infinite recursion checker to not warn if a
branch terminates, but still warns if a branch calls into something with
`@_semantics("arc.programtermination_point")`. This way, calling `fatalError`
doesn't disqualify you for the diagnostic, but calling `exit` does.

This also removes the warning workaround in the standard library, and
annotates the internal _assertionFailure functions as
`programtermination_point`s, so they get this treatment too.

harlanhaskins committed:
[SILOptimizer] Don't diagnose infinite recursion if a branch terminates

This patch augments the infinite recursion checker to not warn if a
branch terminates, but still warns if a branch calls into something with
`@_semantics("arc.programtermination_point")`. This way, calling `fatalError`
doesn't disqualify you for the diagnostic, but calling `exit` does.

This also removes the warning workaround in the standard library, and
annotates the internal _assertionFailure functions as
`programtermination_point`s, so they get this treatment too.

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp:
```diff
@@ -109,6 +111,14 @@ static bool hasInfinitelyRecursiveApply(SILFunction *targetFn) {
+    // Note that this check happens _before_ the recursion check. This is
+    // deliberate, as we want to ignore recursive calls inside a program
+    // termination point.
+    if (isARCInertTrapBB(curBlock))
```
Can you rename this in this commit to is program termination point? Since we are expanding the use of this I would suggest that we also rename the semantics string to something more descriptive since this is no longer used just for ARC.

harlanhaskins replied:
Might as well move it out of ARCAnalysis and into SILBasicBlock, eh?

harlanhaskins replied:
Also, I think we should include something about the lack of side effects in the name, moreso than `isProgramTerminationPoint` would...

gottesmm commented on lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp:
```diff
@@ -119,7 +129,9 @@ static bool hasInfinitelyRecursiveApply(SILFunction *targetFn) {
-    if (curBlock->getTerminator()->isFunctionExiting())
+    TermInst *terminator = curBlock->getTerminator();
+    if (terminator->isFunctionExiting() ||
+        terminator->getTermKind() == TermKind::UnreachableInst)
```
If you are going to go this way can you move this into a separate query function and exhaustive switch over term kind so that if people add a new weird terminator they know to update this?

airspeedswift reviewed (approved):
lgtm for the std lib part

harlanhaskins commented:
@gottesmm I've renamed `isARCInertTrap` to `SILBasicBlock::isProgramTerminationPoint` but I think that name isn't descriptive enough. Thoughts? I've also renamed `arc.programtermination_point` to just `programtermination_point`but I need to go through and rearrange it in the docs to move it out of the 'strictly ARC' section.

gottesmm commented on lib/SIL/SILBasicBlock.cpp:
```diff
@@ -336,6 +337,79 @@ bool SILBasicBlock::isEntry() const {
 }
 
+//===----------------------------------------------------------------------===//
+//                            Leaking BB Analysis
```
Can you change all of the references here to ARC/Leaking BBs/etc to refer to trapping block analysis/etc. I'll show you the ones that I see.

gottesmm commented on lib/SIL/SILBasicBlock.cpp:
```diff
@@ -336,6 +337,79 @@ bool SILBasicBlock::isEntry() const {
+  }
+
+  // Otherwise, we have an unreachable and every instruction is inert from an
+  // ARC perspective in an unreachable BB.
```
Here.

gottesmm commented on lib/SIL/SILBasicBlock.cpp:
```diff
@@ -336,6 +337,79 @@ bool SILBasicBlock::isEntry() const {
+  return false;
+}
+
+/// Match a call to a trap BB with no ARC relevant side effects.
```
Here. Also this shouldn't be a doxygen comment (so 2 slashes).

gottesmm reviewed (approved):
3 nits and then LGTM.

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
(I also want to make sure this doesn't introduce any new warnings)

@swift-ci please test source compatibility

harlanhaskins commented:
So, interestingly, this warning happens in the source compatibility suite, and has been happening for as far back as there are still build logs (before #19693).

```
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/ReactiveSwift/Sources/ValidatingProperty.swift:193:21: warning: all paths through this function will call itself
    public convenience init<U, E>(
```

And I think this is _another_ false positive related to dynamic dispatch, similar to [SR-7925](https://bugs.swift.org/browse/SR-7925)

harlanhaskins commented:
Release source compat run failed while cloning Vapor, but I'm satisfied with the results from the debug run. @gottesmm Good to merge?

gottesmm commented:
Let err rip.

graydon commented:
This is causing failures. See rdar://45080912. Backing out.
